tests/drink/src/test_contract_tests.rs

```diff
@@ -0,0 +1,253 @@
+    let got = session.query(instance.get_array()).unwrap().result.unwrap();
+    assert!(got == [(5, Enum1::A()), (5, Enum1::A())]);
+
+    Ok(())
+}
+
+#[test]
+fn test_compact_messages() -> Result<()> {
+    let mut session: Session<MinimalRuntime> = Session::new().expect("Init new Session");
+    let instance = setup(&mut session, BOB);
+
+    let _r = session
+        .execute(instance.set_compact(scale::Compact(42)))
+        .unwrap();
+    let got = session
+        .query(instance.get_compact())
+        .unwrap()
+        .result
+        .unwrap();
+    assert!(got == scale::Compact(42));
+
+    Ok(())
+}
+
+#[test]
+fn test_messages_with_clashing_argument_names() -> Result<()> {
+    let mut session: Session<MinimalRuntime> = Session::new().expect("Init new Session");
+    let instance = setup(&mut session, BOB);
+
+    let _r = session
+        .execute(instance.set_forbidden_names(1, 2, 3, 4, 5))
+        .unwrap();
+
+    let read = session.query(instance.get_u32()).unwrap();
+    assert!(read.result.unwrap() == 1 + 2 + 3 + 4 + 5);
+
+    Ok(())
+}
+
+#[test]
+fn test_conversion_to_account_id() -> Result<()> {
+    let mut session: Session<MinimalRuntime> = Session::new().expect("Init new Session");
+    let instance = setup(&mut session, BOB);
+
+    let _r = session.execute(instance.set_u32(12345)).unwrap();
+
+    let account_id: AccountId = instance.into();
+    let instance: Instance = account_id.into();
+
+    assert!(session.query(instance.get_u32()).unwrap().result.unwrap() == 12345);
+
+    Ok(())
+}
+
+#[test]
+fn test_events() -> Result<()> {
+    use test_contract::event::Event;
+
+    let mut session: Session<MinimalRuntime> = Session::new().expect("Init new Session");
+    let instance = setup(&mut session, BOB);
+
+    let struct2 = Struct2(
+        Struct1 {
+            a: 1,
+            b: 2,
+            c: [0; 4],
+        },
+        Enum1::B(3),
+    );
+
+    let _r = session
+        .execute(instance.set_u32(123))
+        .unwrap()
+        .result
+        .unwrap();
+
+    let _r = session
+        .execute(instance.set_struct2(struct2.clone()))
+        .unwrap()
+        .result
+        .unwrap();
+
+    let struct1 = session
+        .query(instance.get_struct1())
+        .unwrap()
+        .result
+        .unwrap();
+
+    let txn = session.execute(instance.generate_events()).unwrap();
+
+    let events = ContractEvents::from_iter(&txn.events, instance);
+
+    assert!(
+        events[0]
+            == Ok(Event::Event1 {
+                a: 123,
+                b: struct2.clone(),
+                c: struct1.c,
+                d: (struct1.clone(), struct2),
+                e: Some(struct1),
+            })
+    );
+    assert!(events[1] == Ok(Event::Event2 {}));
+
+    Ok(())
+}
+
+#[test]
+fn test_ink_lang_error() -> Result<()> {
+    let mut session: Session<MinimalRuntime> = Session::new().expect("Init new Session");
+    let instance = setup(&mut session, BOB);
```

Comment:
you can spare many lines in these two test modules by making `setup` return a pair of `Instance` and `Session` :)

Reply:
Done.